storage/src/vespa/storage/bucketdb/bucketdatabase.h

```diff
@@ -76,9 +82,37 @@ class BucketDatabase : public vespalib::Printable
     virtual void forEach(
             EntryProcessor&,
             const document::BucketId& after = document::BucketId()) const = 0;
-    virtual void forEach(
-            MutableEntryProcessor&,
-            const document::BucketId& after = document::BucketId()) = 0;
+
+    struct TrailingInserter {
+        virtual ~TrailingInserter() = default;
+        virtual void insert_at_end(const Entry&) = 0;
+    };
+
+    struct Merger {
+        virtual ~Merger() = default;
+        // Visibility of changes to this object when MergingProcessor::Result::Update
+        // is _not_ returned is undefined.
+        // TODO this should ideally be separated into read/write functions, but this
+        // will suffice for now to avoid too many changes.
+        virtual uint64_t bucket_key() const noexcept = 0;
+        virtual document::BucketId bucket_id() const noexcept = 0;
+        virtual Entry& current_entry() = 0;
+        virtual void insert_before_current(const Entry&) = 0;
+    };
+
+    struct MergingProcessor {
+        enum class Result {
+            Update,
+            KeepUnchanged,
+            Skip
+        };
+
+        virtual ~MergingProcessor() = default;
+        virtual Result merge(Merger&) = 0;
+        virtual void insert_remaining_at_end(TrailingInserter&) {}
+    };
+
+    virtual void merge(MergingProcessor&) = 0;
```

Comment:
Consider adding comment on how an implementation of this is supposed to use the MergingProcessor interface.

Reply:
Done